lib/ipmi_delloem.c

```diff
@@ -338,6 +365,689 @@ usage(void)
+*
+* Description:	This function retrieves tag string
+*
+* Input: 		intf 		- pointer to interface
+*			tag_type - either assettag / servicetag
+*			length	- length of the tagstring
+* Output: 	tagstring - assettag  / servicetag string
+*
+* Return:		0
+*
+******************************************************************/
+
+static int ipmi_sysinfo_get_tag(struct ipmi_intf * intf,
+					char* tagstring, uint8_t tag_type, uint8_t length)
+{
+	struct ipmi_rs * rsp = NULL;
+	struct ipmi_rq req = {0,0,0,0,0,0};
+	uint8_t data[4] = {0};
+	IPMI_DELL_TAG * tagblock;
+
+		memset (&req,0,sizeof(req));
+
+		req.msg.netfn = IPMI_NETFN_APP;
+		req.msg.lun = 0;
+		req.msg.cmd = IPMI_GET_SYS_INFO;
+		req.msg.data_len = 4;
+		req.msg.data = data;
+		data[0] = 0;		/* get parameter*/
+		data[1] = tag_type;	/* get asset tag*/
+		data[2] = 0;
+		data[3] = 0;
+
+		rsp = intf->sendrecv(intf, &req);
+
+		if (rsp == NULL) {
+			lprintf(LOG_ERR, " Error getting asset tag");
+			return -1;
+		} else if (rsp->ccode > 0) {
+			lprintf(LOG_ERR, " Error getting asset tag: %s",
+				val2str(rsp->ccode, completion_code_vals));
+			return -1;
+		}
+
+		tagblock = (IPMI_DELL_TAG *) (void *) rsp->data;
+		memset(tagstring, 0,tagblock->length);
+		memcpy (tagstring, &tagblock->tag, tagblock->length);
+		tagstring[tagblock->length] = '\0' ;
+
+		return 0;
+
+}
+/*****************************************************************
+* Function Name: 	ipmi_sysinfo_print_board
+*
+* Description:	This function prints FRU Board Area for sysinfo command
+* Input: 		intf 			- pointer to ipmi interface
+*			fru			- pointer to fru area
+*			id			- fru id
+*			offset		- offset to fru area
+*			ipmi_version	- ipmi version
+* Output:
+*
+* Return:
+*
+******************************************************************/
+static void
+ipmi_sysinfo_print_board(struct ipmi_intf * intf, struct fru_info * fru,
+		     uint8_t id, uint32_t offset, uint8_t ipmi_version)
+{
+	char * fru_area;
+	uint8_t * fru_data, *fru_data_tmp;
+	uint8_t language;
+	uint32_t fru_len, area_len, i, j;
+	uint16_t length;
+
+    fru_area=NULL;
+	j = 0;
+
+	i = offset;
+	fru_len = 0;
+
+
+	fru_data_tmp = fru_data = malloc(fru->size + 1);
+	if (fru_data == NULL) {
+		lprintf(LOG_ERR, " Out of memory!");
+		return;
+	}
+	memset(fru_data, 0, fru->size + 1);
+
+	/* read enough to check length field */
+	if (read_fru_area(intf, fru, id, i, 2, fru_data_tmp) == 0)
+//		fru_len = 8 * fru_data[i + 1];
+		fru_len = 8 * fru_data[1];
+	if (fru_len <= 0) {
+		free(fru_data);
+		return;
+	}
+
+	fru_len = 0x200;
+	i = 0;
+	/* read in the full fru */
+	if (read_fru_area(intf, fru, id, i, fru_len, fru_data_tmp) < 0) {
+		free(fru_data);
+		return;
+	}
+	BoardOffsetPtr = (fru_data[BOARD_OFFSET])*OFFSETMULTIPLIER;
+	ProductOffsetPtr = (fru_data[PRODUCT_OFFSET])*OFFSETMULTIPLIER;
+	if(iDRAC_FLAG == IDRAC_14G)
+		language = fru_data[BoardOffsetPtr+BOARDLANGOFFSET];
+	else if (iDRAC_FLAG == IDRAC_13G)
+		language = fru_data[10];
```

Comment:
So, iDRAC_13G's FRU is not IPMI-compliant? Or does it simply have Board Area starting at offset 8 ?
Anyway, most of this function's code looks like you're repeating in one way or another other FRU processings functions from `lib/ipmi_fru.c` and `lib/ipmi_ekanalyzer.c`. They already contain a lot of duplicating code and I wouldn't like to add more. Please think of reusing the existing code.

I especially dislike the manual calculation of area and field offsets for standardized areas. This definitely needs not to be reinvented.